src/morgenstern.jl

```diff
@@ -1,108 +1,311 @@
+\\end{aligned}
+```
+
+The parameter ``\\varepsilon \\in \\mathbb{F}_q`` is chosen so that the polynomial ``f(x) = x^2 + x + \\varepsilon``
+is irreducible over ``\\mathbb{F}_q``. This ensures ``\\mathcal{A}`` is a skewfield. The algebra is [ramified](https://en.wikipedia.org/wiki/Ramification_(mathematics))
+at the finite place ``x`` and at ``1/x``.
+
+The connection to graph theory arises from studying elements in the *integral set*
+
+```math
+\\begin{aligned}
+\\mathscr{S} = \\mathbb{F}_q[x]\\mathbf{1} + \\mathbb{F}_q[x]\\mathbf{i} + \\mathbb{F}_q[x]\\mathbf{j} + \\mathbb{F}_q[x]\\mathbf{ij}
+\\end{aligned}
+```
 
-Takes as input 𝔽q[x], the polynomial ring we want to work with.
+The "basic norm x+1" elements are defined as
 
-Returns (ε, sols), where sols is the list of solutions.
+```math
+\\begin{aligned}
+\\xi = 1 + \\gamma\\mathbf{j} + \\delta\\mathbf{ij}, \\quad \\text{with} \\gamma, \\delta \\in \\mathbb{F}_q,
+\\end{aligned}
+```
 
-See [morgenstern1994existence](@cite).
+satisfying the norm equation ``N(\\xi) = \\gamma^2 + \\gamma\\delta + \\delta^2\varepsilon = 1``. This equation
+has exactly ``q+1`` solutions in ``\\mathbb{F}_q``, parameterizing generators ``\\xi_1, \\dots, \\xi_{q+1}``. These
+generators define a [free product](https://en.wikipedia.org/wiki/Free_product) group 
+
+```math
+\\begin{aligned}
+A(x) = \\langle \\xi_1 \\rangle * \\langle \\xi_2 \\rangle * \\cdots * \\langle \\xi_{q+1} \\rangle
+\\end{aligned}
+```
+
+which acts simply transitively on the ``q+1``-regular tree ``T_{x+1} = G'_{x+1}/G'_{O_{x+1}}``. Taking the
+quotient by a [congruence subgroup](https://en.wikipedia.org/wiki/Congruence_subgroup) ``A(g)``, where ``g(x)``
+is irreducible of even degree, yields a finite ``(q+1)``-regular graph ``\\Gamma_g = A(g) \\backslash T_{x+1}``.
+This graph is the Cayley graph of ``PSL_2(\\mathbb{F}_{q^d})`` with respect to the images of the ``q+1`` generators.
+
+# Arguments
+- `R`: Polynomial ring ``\\mathbb{F}_q[x]`` where `q` is a power of 2.
 """
-function morgenstern_solutions(R)
+function morgenstern_solutions(R::FqPolyRing)
     F = base_ring(R)
     unit = gen(F)
     q = order(F)
     f = morgenstern_f(R) # random sampler
-    ε = coefficients(f)[0]
+    ε = coeff(f, 0)
+    # (1, 0) is always a solution: 1² + 1·0 + 0²·ε = 1
     sols = [(one(F),zero(F))]
-    for s in F
-        fs = f(s)
-        sfs = sqrt(fs)
-        α = s*inv(sfs)
-        @assert s*inv(sfs) == inv(sfs)*s
-        β = inv(sfs)
-        @assert α^2 + α*β + ε*β^2 == 1
-        push!(sols, (α,β))
+    # Theorem 5.13 of [morgenstern1994existence](@cite): find all solutions (γ, δ) ∈ 𝔽q² to γ² + γδ + δ²ε = 1
+    for δ in F
+        iszero(δ) && continue
+        for γ in F
+            if γ^2+γ*δ+δ^2*ε == one(F)
+                push!(sols, (γ, δ))
+                # in GF2, if (γ, δ) is solution, so is (γ+δ, δ)
+                other_γ = γ+δ
+                push!(sols, (other_γ, δ))
+                break
+            end
+        end
```

Comment:
Check performance -- you are turning a (probably) linear algorithm into a quadratic one.

On the other hand, this new implementation is much easier to understand.

If it is slower, remove it from here and use it as a test.

If it is faster, keep this new implementation.

Make sure you check performance with a large F (large q) so that we are not looking only at very small solutions where constant factors dominate the complexity.